crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,832 @@
+    let q_target = Qubit(0);
+    let q_controls: Vec<Qubit> = (1..num_qubits).map(|i| Qubit(i as u32)).collect();
+    decompose_uc_rotations(angles, 0, angles.len(), false);
+    for (i, angle) in angles.iter().enumerate() {
+        if angle.abs() > EPS {
+            let _ = out.push_standard_gate(StandardGate::RZ, &[Param::Float(*angle)], &[q_target]);
+        }
+        if i != angles.len() - 1 {
+            // Is this a correct translation from python?
+            // binary_rep = np.binary_repr(i + 1)
+            // q_contr_index = len(binary_rep) - len(binary_rep.rstrip("0"))
+            let q_ctrl_index = (i + 1).trailing_zeros();
+            let _ = out.push_standard_gate(
+                StandardGate::CX,
+                &[],
+                &[q_controls[q_ctrl_index as usize], q_target],
+            );
+        } else if vw_type_all {
+            let q_ctrl_index = num_qubits - 2;
+            let _ = out.push_standard_gate(
+                StandardGate::CX,
+                &[],
+                &[q_controls[q_ctrl_index], q_target],
+            );
+        };
+    }
+    Ok(out)
+}
+
+/// Calculates rotation angles for a uniformly controlled R_t gate with a C-NOT gate at
+/// the end of the circuit. The rotation angles of the gate R_t are stored in
+/// angles[start_index:end_index]. If reversed_dec == True, it decomposes the gate such that
+/// there is a C-NOT gate at the start of the circuit (in fact, the circuit topology for
+/// the reversed decomposition is the reversed one of the original decomposition)
+fn decompose_uc_rotations(
+    angles: &mut [f64],
+    start_index: usize,
+    end_index: usize,
+    reversed_decomposition: bool,
+) {
+    let interval_len_half = (end_index - start_index) / 2;
+    for i in start_index..start_index + interval_len_half {
+        if !reversed_decomposition {
+            let new_angles = update_angle(angles[i], angles[i + interval_len_half]);
+            angles[i] = new_angles[0];
+            angles[i + interval_len_half] = new_angles[1];
+        } else {
+            let new_angles = update_angle(angles[i], angles[i + interval_len_half]);
+            angles[i + interval_len_half] = new_angles[0];
+            angles[i] = new_angles[1];
+        }
+    }
+    if interval_len_half > 1 {
+        decompose_uc_rotations(angles, start_index, start_index + interval_len_half, false);
+        decompose_uc_rotations(angles, start_index + interval_len_half, end_index, true);
+    }
+}
+
+fn update_angle(angle_1: f64, angle_2: f64) -> [f64; 2] {
+    [(angle_1 + angle_2) / 2., (angle_1 - angle_2) / 2.]
+}
+
+fn append(circ: &mut CircuitData, new: CircuitData, qubit_map: &[Qubit]) -> PyResult<()> {
+    for inst in new.data() {
+        let qubits_map: Vec<Qubit> = new
+            .get_qargs(inst.qubits)
+            .iter()
+            .map(|x| qubit_map[x.index()])
+            .collect();
```

Comment:
I think there is a path to avoid the allocation here. @jakelishman might have a pattern off the top of his head, but allocating a vec on each allocation seems unnecessary for this.

Reply:
I guess we can always use a dict to store pre-encountered mapped qubits:

```
let mut new_qubits_map: HashMap<Interned<[Qubit]>, Vec<Qubit>> = HashMap::new();
    for inst in new.data() {
        let qubits_map = new_qubits_map.entry(inst.qubits).or_insert_with(|| { 
            new
            .get_qargs(inst.qubits)
            .iter()
            .map(|x| qubit_map[x.index()])
            .collect()
        });
```

Do you think this can be further optimized?